design/ceph/ceph-orchestrator-osd-add-remove.md

```diff
@@ -0,0 +1,169 @@
+# Adding and removing Ceph osds via Ceph orchestrator interface
+
+## Goals
+
+1. Give Ceph's Rook Orchestrator Plugin the ability to add and remove OSDs to/from a Rook-Ceph
+   cluster running in Kubernetes.
+2. Users of the Ceph Orchestrator CLI and/or Ceph Dashboard should be able to
+   create OSDs on nodes with as much complexity as Ceph is able.
+3. Do not introduce unnecessary complexity into Rook or into Rook's CephCluster CRD.
+
+## Proposal
+
+### Removing OSDs
+Removing OSDs is the simplest place to start with this design. When removing an OSD or OSDs via the
+Rook Orchestrator Module, the Ceph mgr running in Kubernetes will run a Kubernetes Job that will run
+the required `ceph-volume` command necessary to remove the OSD(s) from a node as desired by the Ceph
+user. The Orchestrator Module will also perform whatever Ceph operations are necessary to remove the
+OSD as would be expected by the user. This will replace the manual, imperative steps currently
+required by Rook-Ceph users to remove OSDs from the storage cluster.
+
+#### OSD removal in the CephCluster resource
+The CephCluster CRD allows adding disks to nodes in several places including disks on all nodes in
+the Kubernetes cluster, to a specific node, or via disk pattern matching (in two different ways). To
+avoid the complexity of having to manage the logic for determining how to remove a node from the
+CephCluster resource, OSD removal will **not** alter the CephCluster CRD. While doing so would be
+convenient for users, there are a lot of opportunities in development to miss corner cases or
+introduce logic bugs. This may mean that Rook could re-add the disk after the disk is removed via
+the Ceph Orchestrator interface if the CephCluster CRD is configured to do so, and the user should
+be made aware of this possibility via documentation.
+
+### Adding OSDs
+When adding an OSD or OSDs via the Rook Orchestrator Module, the Ceph mgr running in Kubernetes will
+run a Kubernetes Job that will run the required `ceph-volume` command necessary to provision the
+OSD(s) on a node node as desired by the Ceph user. This will be an imperative operation and will not
+affect the CephCluster resource. Based on arguments from
+[OSD removal in the CephCluster resource](#osd-removal-in-the-cephcluster-resource), adding OSDs via
+the Ceph orchestrator interface will not add disks to the CephCluster resource so as to maintain
+internal consistency between add and remove operations.
+
+Additionally, if OSDs were to be added via the CephCluster resource, in order to maintain goal
+number 2, the CephCluster CRD would have to be modified to either:
+
+1. Add a full understanding of `ceph-volume`'s options including Drive Groups, OR
```

Comment:
Supporting all (reasonable) ceph-volume options from the CRD doesn't seem that complex to me. Why wouldn't we add support for drive groups into the CR? We already have many different ways to create OSDs under the `storage` element such as raw devices, partitions, or the `storageClassDeviceSet`.  Adding another section for drive groups doesn't seem out of the question. But I need to review drive groups again so I might be oversimplifying.

Reply:
What I **don't** want is for the drive group specification to change over time and for both Rook and Ceph to have to maintain internal knowledge of what drive group spec is supported by which Rook and Ceph version. We can avoid this, but we need to make sure the drive group spec is versioned. And honestly, we might want the drive group spec in Rook to be a dumb passthrough and to not have Rook internalize the spec whatsoever.

Reply:
Also, the word `reasonable` is tricky. I'm sure both our organizations have users who want to do things that we might think aren't reasonable but which to them are "requirements". I don't want to have to maintain the internal knowledge of `ceph-version` versions within Rook to maintain an idea of which flags are supported and how features map to flags of "reasonable" options. It would be much easier to have the CRD pass through arbitrary flags to `ceph-volume` with a drive group spec or to just specify a whole `ceph-volume` command with whatever flags it desires. This will be hugely complicated for users without very deep Ceph knowledge, but I think our intention should be that this method is really used only by Ceph's Rook Orchestrator Module itself.

Reply:
@BlaineEXE @sebastian-philipp After a quick review of the [drive groups design](https://github.com/SUSE/DeepSea/wiki/Drive-Groups), my impression is:
1. In an ideal world where we don't have to worry about rook vs ceph versioning I would like to see the drive groups translated into the CR.
1. Drive groups have a lot of overlapping properties with settings we already have in the CR: encryption, wal size, device name, etc. For properties already implemented in rook, there could be a simple translation to the cluster CR settings.
1. The drive groups have useful new functionality to filter based on various device properties such as size, serial number, or manufacturer. Ideally these properties could also be implemented in the cluster CR. It is a common request to add properties like this to the CR.
1. In the real world, there is a non-trivial versioning matrix that would make it complex to translate the drive groups into the CR and would make it difficult to support a changing drivegroups spec. 

My thoughts are now to recommend the pass-through approach as discussed today. Given that the mgr modules and ceph-volume will be versioned together in the Ceph repo, the versioning challenge goes away if we pass through the drive spec to the cluster CR. Specifically, this means:
- The mgr module would add the drivegroup spec to the cluster CR
- The operator would start an orchestration, which would pass the drivegroup spec for a node to the osd-prepare pod, which would call c-v with the drivegroup spec. c-v would handle all provisioning of the drive group with the single command.
- Neither the mgr module nor rook should translate the drivegroup spec into c-v commands. We need a single c-v command that consumes the drivegroup. Otherwise, the versioning issues would appear again.
- Orchestrator modules wouldn't need to do any processing of drive groups. They should just pass them on to the orchestrator to pass through the spec to c-v on a node.

But I have the impression that the mgr module is doing some translation to c-v commands. Can we move this processing into c-v? 

One aspect of drive groups that the operator would need to process is what nodes the settings apply to. The operator needs to decide on which nodes to start the osd prepare pods. 

The drive groups in the CR could also be useful as a new way for CR users to create OSDs. Instead of using properties such as `useAllDevices` or `deviceFilter`, they could use drivegroups where new properties are needed. The drive groups don't look that complex at least in the link above. We could add the drive groups to the CR and make it look just like other properties, except that internally we would just deserialize it and pass it through as a blob.

Reply:
cc @jan--f. You thoughts?

Reply:
To include the drive group specification directly in the CRD can have undesired collateral effects. We should remember that drive groups is just a way to describe a device selection made in one point of the time. If the underlying infrastructure or the user needs changes then the result is not valid.

The drive group specification is not stored in any place (dashboard, orchestrator), is something that is used only one time to do a specific operation over a specific set of devices in one moment. ( another point to not include this in the CRD)

I explained my point of view about to have an "abstract specification" of the OSDs in the cluster CRD, ( in previous comment about the use of useAllNodes, useAllDevices... ), this produces a vision of the cluster with undesired collateral effects.

I should insist that probably the best way to deal with this is to have in the cluster CRD the complete specification for each OSD.
One posible idea of implementation would be:
The dashboard/orchestrator cli is going to use a "drive group" to specify a creation of OSD's. The Rook orchestrator should patch the cluster CRD with the ceph-volume command, and the OSD's attributes of each OSD included in the drive group.
Rook should the thing that it better do, reconcile the real cluster status with the cluster CRD, creating the OSD's using the provide ceph-volume orders and creating the needed pods.